src/System.Management.Automation/namespaces/FileSystemProvider.cs

```diff
@@ -31,6 +31,72 @@
 
 namespace Microsoft.PowerShell.Commands
 {
+    #region InodeTracker
+    /// <summary>
+    /// Tracks visited files/directories by caching their device IDs and inodes.
+    /// </summary>
+    internal class InodeTracker
+    {
+        private Dictionary<UInt64, Dictionary<UInt64, bool>> _visitations;
+
+        /// <summary>
+        /// Construct a new InodeTracker
+        /// </summary>
+        public InodeTracker()
+        {
+            _visitations = new Dictionary<UInt64, Dictionary<UInt64, bool>>();
+        }
+
+        /// <summary>
+        /// Determine if a path has been visited.
+        /// </summary>
+        /// <param name="path">Path to the file or directory to be checked.</param>
+        /// <returns>True if the path has been visited, false otherwise</returns>
+        public bool Visited(string path)
+        {
+            UInt64 dev;
+            UInt64 inode;
+
+            if (InternalSymbolicLinkLinkCodeMethods.GetInodeData(path, out dev, out inode))
+            {
+                if (_visitations.ContainsKey(dev))
+                {
+                    return _visitations[dev].ContainsKey(inode);
```

Comment:
Part of that discussion ([https://stackoverflow.com/questions/2877660/composite-key-dictionary#comment14110652_2877820](https://stackoverflow.com/questions/2877660/composite-key-dictionary#comment14110652_2877820)) suggests that the nested dictionary approach may be faster.

As for the C#7 pattern, I thought I read somewhere in another code review that C#7 code should be avoided. Is that not the case?